### DISCUSSION TITLE
General

### DISCUSSION THREAD

**Discussion by Infinite-Coder-001** [Infinite-Coder-001/Hotlist-API#1] [General]
# General
**Welcome to Hotlist API general discussion thread!** 
In this thread, you can ask or comment something or report a bug.

**Comment by Reginald-Gillespie** [comment]
Moving to discussions 😎

Yeah so google doesn't seem to have recaptcha, but I'd bet after you make a few accounts it will put you through hCaptcha

**Reply by Infinite-Coder-001** [reply]
Noice. True, Google is proffesional, it 100% has a protection. 

So that's propably all. Thanks for help and see you maybe next time 🙂.

**Reply by Reginald-Gillespie** [reply]
Lol yw, I have fun with API stuff, but KA always finds my accounts before I have time to get it out there well enough, so it was nice to see you release it because now it will last.

**Reply by Reginald-Gillespie** [reply]
I see you're working on the Sketch API, do you have an idea for how it's gonna pull in which programs it should update info for?

**Reply by Infinite-Coder-001** [reply]
I'll make a thread in Tips&Thanks in the program, that I'll release on Khan Academy. If somebody wants to use this API, he must send his program id into the thread. Then, the updating script will load all comments with ids and update info for it. But I propably can't do this on GAS, because it has the limit of requests. So I'll do it on Replit.

**Reply by Infinite-Coder-001** [reply]
Fetching the data already works, I need to connect to Github now.

**Reply by Reginald-Gillespie** [reply]
Having them tell you is the best (and only practical) way. The only way to send data outside KA is to use a custom turn server, here's what I told someone else on KA:
```
In case you are wondering how to do it though from KA, here's how (I'm fairly certain this is the only way):
UDP is unrestricted by cors, so you use WebRTC to connect to a turn server that you own. You send the link to the code in the username field, and on the turn server you run a custom debug log function that extracts the link from the username field. Then you request the program from the KA API, format the code, and push it in a string form to a github repo named with the program ID. Then from the PJS program, you keep attempting to fetch the github repo until it's there. Then display the results to them.

Hosting the turn server requires that you have a device or VPS with a public IP (so port forwarded) so that it can be accessible from anywhere on the internet.

My guess is that you don't have the hardware to run the turn server, so the best course of action is to give up the idea =P
```

**Reply by Infinite-Coder-001** [reply]
I'm not trying to send data somewhere from a KA sketch. I'm just trying to do something similiar to Hotlist API, but it will get program code, forks, etc.

**Reply by Reginald-Gillespie** [reply]
I just heard from HSstudent that webworkers are not bound by cors... I haven't looked into it myself, but it sounds like a possible vector to renew the API easier than self-hosting updaters like we've done so far.

**Reply by Infinite-Coder-001** [reply]
I tried it, but it shows this error message (in Edge): 

Uncaught DOMException: Failed to construct 'Worker': Script at 'https://www.w3schools.com/html/demo_workers.js' cannot be accessed from origin 'https://www.kasandbox.org'.

I don't know how workers could help me with the API. 

And thanks for re-subscribing by the way 😃

I was thinking about the auto-updated images. There's really no way around tainting the canvas but if someone wanted to do something like a live leaderboard, he could edit the updating script so the whole leaderboard will be updated and uploaded to KA as an image.

**Reply by Reginald-Gillespie** [reply]
HSstudent said there were some more problems with it but not sure how far those go. So might not be possible, I'll have to try

Well, yeah you'd have to initialize it from a string, not a url. 

They issue we're dealing with is cors, if we can get past that then we can access the api from a scratchpad dirrectly.

**Comment by Reginald-Gillespie** [comment]
https://www.khanacademy.org/python-program/pyfetch-v1/5482075651751936

Because their new python is run inside a worker, we can now use the API again (with a little bit of CORs bypass).
